// jet-tools/lib/enums/measurement-unit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeasurementUnit = void 0;
var MeasurementUnit;
(function (MeasurementUnit) {
    MeasurementUnit["oz"] = "oz";
    MeasurementUnit["ml"] = "ml";
    MeasurementUnit["cl"] = "cl";
    MeasurementUnit["part"] = "part";
    MeasurementUnit["tbsp"] = "tbsp";
    MeasurementUnit["tsp"] = "tsp";
    MeasurementUnit["dash"] = "dash";
    MeasurementUnit["drop"] = "drop";
    MeasurementUnit["barspoon"] = "barspoon";
    MeasurementUnit["onTop"] = "onTop";
    MeasurementUnit["cup"] = "cup";
    MeasurementUnit["piece"] = "piece";
    MeasurementUnit["wedge"] = "wedge";
    MeasurementUnit["slice"] = "slice";
    MeasurementUnit["coin"] = "coin";
    MeasurementUnit["item"] = "item";
    MeasurementUnit["cube"] = "cube";
    MeasurementUnit["leaf"] = "leaf";
    MeasurementUnit["sprig"] = "sprig";
    MeasurementUnit["pinch"] = "pinch";
    MeasurementUnit["scoop"] = "scoop";
})(MeasurementUnit = exports.MeasurementUnit || (exports.MeasurementUnit = {}));
